dbms/src/Operators/SharedAggregateRestorer.cpp

```diff
@@ -66,7 +82,7 @@ void SharedSpilledBucketDataLoader::storeBucketData()
     RUNTIME_CHECK(status == SharedLoaderStatus::loading);
 
     // Although the status will always stay at `SharedLoaderStatus::loading`, but because the query has stopped, it doesn't matter.
-    if unlikely (exec_context.isCancelled())
+    if (checkCancelled())
```

Comment:
Why `SharedQuery` and `ResultQueue` need to be cancelled explicitly in `PipelineExecutorContext::cancel`, but this don't need to be cancelled in `PipelineExecutorContext::cancel`?

Reply:
Because it will not be stuck forever, when the query is canceled, checkCancel will always be called, because wait_for_notify must be because the queue is empty. When the queue is empty, it will trigger an asynchronous load bucket event, and the load bucket event will call checkCancel.

Reply:
If ` trigger an asynchronous load bucket event` fails, will it still work?

Reply:
yes, `Event::finishImpl` is always called, regardless of failure or cancellation
https://github.com/pingcap/tiflash/blob/a4f0ec6ea19659062e2596c772a9bfcfab115c07/dbms/src/Flash/Pipeline/Schedule/Events/Impls/LoadBucketEvent.cpp#L29-L33

Reply:
https://github.com/pingcap/tiflash/pull/9082/files#diff-b54d5b998f3a9de97d9452c7c40a8016e79ecbc2251784e96ac0fd6ee840fa16R119-R120

Reply:
OK. I want to confirm another unrelated question here, why `WorkQueue` in this [pr](https://github.com/pingcap/tiflash/pull/8872/files) don't need to be cancelled explicitly in `PipelineExecutorContext::cancel`?

Reply:
When the data in the storage layer is fully read, notify will definitely be called.
When the data in the storage layer is not fully read, there will always be a block, and notify will also be called.

Reply:
When query meet error, the storage layer is not informed?

Reply:
Yes, the storage layer is not explicitly canceled.

Reply:
However, the query will not be stuck, because the storage layer will continuously generate blocks, which will trigger checkCancel, thus ending the query.

Reply:
Storage layer generate block => pipeline task is notified and finishes since it found the query is cancelled => `~UnorderedSourceOp()` is called, and in side `~UnorderedSourceOp()`, it will call `task_pool->decreaseUnorderedInputStreamRefCount()`, which eventually stop the storage read?

Reply:
> Storage layer generate block => pipeline task is notified and finishes since it found the query is cancelled => `~UnorderedSourceOp()` is called, and in side `~UnorderedSourceOp()`, it will call `task_pool->decreaseUnorderedInputStreamRefCount()`, which eventually stop the storage read?

yes